src/lua/LuaSerializer.cpp

```diff
@@ -359,12 +367,85 @@ void LuaSerializer::UninitTableRefs()
 	lua_setfield(l, LUA_REGISTRYINDEX, "PiSerializer");
 
 	lua_pushnil(l);
-	lua_setfield(l, LUA_REGISTRYINDEX, "PiSerializerTableRefs");
+	lua_setfield(l, LUA_REGISTRYINDEX, NS_REFTABLE);
 
 	lua_pushnil(l);
 	lua_setfield(l, LUA_REGISTRYINDEX, "PiLuaRefLoadTable");
 }
 
+void LuaSerializer::SavePersistent(Json &json)
+{
+	lua_State *l = Lua::manager->GetLuaState();
+	LUA_DEBUG_START(l);
+
+	// NOTE: this must be an array for consistent deserialization order
+	Json persist = Json::array();
+
+	luaL_getsubtable(l, LUA_REGISTRYINDEX, NS_PERSISTENT);
+	lua_pushnil(l);
+
+	while (lua_next(l, -2)) {
+		Json entry = Json::object();
+
+		// persistent, id, value
+		std::string id = LuaPull<std::string>(l, -2);
+
+		// Serialize all registered persistent objects as a "backup" version
+		// (This will populate the PiSerializerTableRefs cache)
+		entry["persistId"] = id;
+		pickle_json(l, -1, entry, id);
+
+		persist.push_back(std::move(entry));
+
+		lua_pop(l, 1);
+	}
+
+	lua_pop(l, 1);
+
+	json["lua_persistent_json"] = std::move(persist);
+
+	LUA_DEBUG_END(l, 0);
+}
+
+void LuaSerializer::LoadPersistent(const Json &json)
+{
+	lua_State *l = Lua::manager->GetLuaState();
+	LUA_DEBUG_START(l);
+
+	const Json &persist = json["lua_persistent_json"];
+
+	luaL_getsubtable(l, LUA_REGISTRYINDEX, NS_REFTABLE);
+	int idx_reftable = lua_gettop(l);
+
+	luaL_getsubtable(l, LUA_REGISTRYINDEX, NS_PERSISTENT);
+
+	for (const auto &item : persist) {
+		std::string id = item["persistId"].get<std::string>();
+
+		lua_pushinteger(l, item["ref"].get<lua_Integer>());
+		lua_getfield(l, -2, id.c_str()); // reftable, persist, ptr, pvalue
+
+		// We must unpickle the saved persistent object, because it may contain
+		// "trapped" serialized representations of non-persistent tables
+		unpickle_json(l, item);
```

Comment:
Taking this opportunity, could you please document a little about the side effects and the non-obvious result of this function?

Reply:
Heh, returning the favor I see. Sure, I can definitely do that :smile: